### PR TITLE
RFC: Import python-fp17 history via subtree merge

### PR DESCRIPTION
**Pull request by lamby**
Odonto will clearly require the code that is currently in the `python-fp17` repo in some hand-wavey way. I think we have two options here:

a) Pull in the code (retaining history) as demonstrated in this PR, although it would clearly need some additional path/module fixups (!) and possibly even entangle it much closer into the Odonto code etc. where necessary. This will help us move quickly and, to be honest, we were going to be the only consumers of the library anyway. We can certainly keep "stuff" reasonably separate just as if we were coding in a sane way, of course.

b) Use a Git submodule. I'm not _personally_ a big fan of these and as discussed previously it would imply that the Python FP17 library is pretty standalone when, in practice, it would kinda not be.

Thoughts? (Please excuse rambling)

### CODE REVIEW

**Comment by fredkingham** [comment]
looks good.
I added the requirements to the odonto requirements, as that seems to make sense.
Are we keeping the requirements/gitignore/travis so that we can easily split it out later?

I'm guessing when travis runs we need to make sure it also runs on pytest for this.

**Comment by lamby** [comment]
@fredkingham Did you see this was just one of two alternatve implementations? Your "looks good", in this context, makes me think I didn't communicate this very well.

**Comment by fredkingham** [comment]
Sorry, I saw you weren't a big fan and somehow read it that these were the two options and you were shooting with a. My bad.

I've never worked with submodules before, I contemplated using them a few years back and then rejected it as unnecessary overhead.

As you say, no one is due to use the module. I'm also not sure how useful it is independent of state.

I'd still keep modularised, I think when we're tracking state, the question of whether we would put that state together with the fp 17 logic or separately but related is still open.

TLDR I agree, option a is the way to go.

**Comment by davidmiller** [comment]
FWIW: 

Every time I've gone near git submodules used in anger, they have frequently made people trying to use them angry... :)

I don't think the separation of the library is that useful right now - I definitely wouldn't want it at the expense of development speed if that seems like the trade off...

**Comment by lamby** [comment]
> read it that these were the two options and you were shooting with a. My bad.

Well, that's where my gut was leading me at least but I was mostly soliciting your fine comments, insight and experience... as well as very slightly rubber-ducking the whole thing. :)

> Are we keeping the requirements/gitignore/travis so that we can easily split it out later?

So, I was planning on merging the functionality of these files into the current `odonto` ones so that the module continues to be tested, otherwise that would be a regression of sorts. To keep the literal files would be somewhat misleading in that they could easily get out of sync. Thoughts?

> […]

ACK re. submodule experience. I'll go ahead and merge this style in some description (ie. retaining history etc. etc.), but not sure exactly how that would look atm. Will leave this open for the moment but pretty much end up merging something like this once it is reasonably presentable.
